// @robinl/to-embed-in-splink-outputs/5c0f40059131a129@520.js
// https://observablehq.com/@robinl/to-embed-in-splink-outputs@520
export default function define(runtime, observer) {
  const main = runtime.module();
  main.variable(observer()).define(["md"], function(md){return(
md`# To embed in splink outputs`
)});
  main.variable(observer("viewof selected_cluster_id")).define("viewof selected_cluster_id", ["splink_vis_utils","cluster_unique_ids"], function(splink_vis_utils,cluster_unique_ids){return(
splink_vis_utils.select(cluster_unique_ids, {
  label: "Choose cluster"
})
)});
  main.variable(observer("selected_cluster_id")).define("selected_cluster_id", ["Generators", "viewof selected_cluster_id"], (G, _) => G.input(_));
  main.variable(observer("nothing_selected_message")).define("nothing_selected_message", ["no_edge_selected","no_node_selected","html","selected_edge"], function(no_edge_selected,no_node_selected,html,selected_edge)
{
  if (no_edge_selected && no_node_selected) {
    return html`<span style="color:red"'>Click on nodes and/or edges in the below graph to show data and chart</span>`;
  }

  if (typeof selected_edge == 'undefined') {
    return html`<span style="color:red"'>Click on an edges in the below graph to show waterfall chart and table</span>`;
  }

  return html``;
}
);
  main.variable(observer("viewof force_directed_chart")).define("viewof force_directed_chart", ["vegaEmbed","splink_vis_utils","spec"], function(vegaEmbed,splink_vis_utils,spec){return(
vegaEmbed(splink_vis_utils.cloneDeep(spec))
)});
  main.variable(observer("force_directed_chart")).define("force_directed_chart", ["Generators", "viewof force_directed_chart"], (G, _) => G.input(_));
  main.variable(observer("edge_table_header")).define("edge_table_header", ["no_edge_selected","html"], function(no_edge_selected,html)
{
  if (no_edge_selected) {
    return html``;
  }
  return html`<h3>Rows compared by selected edge</h3>`;
}
);
  main.variable(observer("edge_table")).define("edge_table", ["no_edge_selected","html","splink_vis_utils","selected_edge","ss"], function(no_edge_selected,html,splink_vis_utils,selected_edge,ss)
{
  if (no_edge_selected) {
    return html``;
  }
  return splink_vis_utils.edge_row_to_table(selected_edge, ss);
}
);
  main.variable(observer("waterfall_chart")).define("waterfall_chart", ["no_edge_selected","html","splink_vis_utils","selected_edge","ss","vegaEmbed"], function(no_edge_selected,html,splink_vis_utils,selected_edge,ss,vegaEmbed)
{
  if (no_edge_selected) {
    return html``;
  } else {
    let waterfall_data = splink_vis_utils.get_waterfall_chart_data(
      selected_edge,
      ss
    );

    return vegaEmbed(
      splink_vis_utils.get_waterfall_chart_spec(waterfall_data, {})
    );
  }
}
);
  main.variable(observer("node_table_header")).define("node_table_header", ["no_node_selected","html"], function(no_node_selected,html)
{
  if (no_node_selected) {
    return html``;
  }
  return html`<h3>Table of selected node</h3>`;
}
);
  main.variable(observer("node_table")).define("node_table", ["no_node_selected","html","splink_vis_utils","selected_node"], function(no_node_selected,html,splink_vis_utils,selected_node)
{
  if (no_node_selected) {
    return html``;
  }
  return splink_vis_utils.node_row_to_table(selected_node);
}
);
  main.variable(observer()).define(["ss"], function(ss){return(
ss.cols_used_by_model()
)});
  main.variable(observer("spec")).define("spec", ["splink_vis_utils","raw_nodes_data","svu_options","selected_cluster_id","ss","raw_edges_data"], function(splink_vis_utils,raw_nodes_data,svu_options,selected_cluster_id,ss,raw_edges_data)
{
  let filtered_nodes = splink_vis_utils.filter_nodes_with_cluster_id(
    raw_nodes_data,
    svu_options.cluster_colname,
    selected_cluster_id
  );

  let formatted_nodes = splink_vis_utils.format_nodes_data_for_force_directed(
    filtered_nodes,
    ss
  );

  let filtered_edges = splink_vis_utils.filter_edges_with_cluster_id(
    raw_edges_data,
    svu_options.cluster_colname,
    selected_cluster_id
  );

  let formatted_edges = splink_vis_utils.format_edges_data_for_force_directed(
    filtered_edges,
    ss
  );
  let s = new splink_vis_utils.ForceDirectedChart(
    formatted_nodes,
    formatted_edges
  );

  s.remove_all_sliders();

  return s.spec;
}
);
  main.variable(observer("ss")).define("ss", ["splink_vis_utils","splink_settings"], function(splink_vis_utils,splink_settings){return(
new splink_vis_utils.SplinkSettings(JSON.stringify(splink_settings))
)});
  main.variable(observer("no_edge_selected")).define("no_edge_selected", ["selected_edge"], function(selected_edge){return(
typeof selected_edge == 'undefined'
)});
  main.variable(observer("no_node_selected")).define("no_node_selected", ["selected_node"], function(selected_node){return(
typeof selected_node == 'undefined'
)});
  main.variable(observer()).define(["md"], function(md){return(
md`## Data processing`
)});
  main.variable(observer("cluster_unique_ids")).define("cluster_unique_ids", ["splink_vis_utils","raw_nodes_data"], function(splink_vis_utils,raw_nodes_data){return(
splink_vis_utils.get_unique_cluster_ids_from_nodes_data(
  raw_nodes_data,
  'cluster_medium'
)
)});
  main.variable(observer("selected_node")).define("selected_node", ["observe_chart_data","force_directed_chart"], function(observe_chart_data,force_directed_chart){return(
observe_chart_data(force_directed_chart, "node_click")
)});
  main.variable(observer("selected_edge")).define("selected_edge", ["observe_chart_data","force_directed_chart"], function(observe_chart_data,force_directed_chart){return(
observe_chart_data(force_directed_chart, "edge_click")
)});
  main.variable(observer("sel_edge")).define("sel_edge", function(){return(
{
  tf_adjusted_match_prob: 0.9999017555,
  match_probability: 0.994335954,
  source_dataset_l: "df_2",
  unique_id_l: 249,
  source_dataset_r: "df_2",
  unique_id_r: 250,
  first_name_l: "Freddi ",
  first_name_r: null,
  gamma_first_name: -1,
  surname_l: "Lewis",
  surname_r: "Lewis",
  gamma_surname: 2,
  dob_l: "1971-12-19",
  dob_r: "1971-12-19",
  gamma_dob: 1,
  email_l: "tyler28@wecveralvarez.aom",
  email_r: "tyler28@weaver-alvarez.com",
  gamma_email: 0,
  group_l: 63,
  group_r: 63,
  cluster_medium_l: 3,
  cluster_medium_r: 3
}
)});
  main.variable(observer("observe_chart_data")).define("observe_chart_data", ["Generators"], function(Generators){return(
function observe_chart_data(chart, signal_name) {
  return Generators.observe(function(notify) {
    // change is a function; calling change triggers the resolution of the current promise with the passed value.

    // Yield the element’s initial value.
    const signaled = (name, value) => notify(chart.signal(signal_name));
    chart.addSignalListener(signal_name, signaled);
    notify(chart.signal(signal_name));

    return () => chart.removeSignalListener(signal_name, signaled);
  });
}
)});
  main.variable(observer()).define(["md"], function(md){return(
md`## Following are global variables embedded in final html so not needed in final version`
)});
  main.variable(observer("localUrl")).define("localUrl", function(){return(
"http://127.0.0.1:8080/dist/splink_vis_utils.js"
)});
  main.variable(observer("splink_vis_utils")).define("splink_vis_utils", ["require","localUrl"], function(require,localUrl){return(
require(`${localUrl}?${Date.now()}`)
)});
  main.variable(observer("inputs")).define("inputs", ["require"], function(require){return(
require("@observablehq/inputs@0.8.0")
)});
  main.variable(observer("vegaEmbed")).define("vegaEmbed", ["require"], function(require){return(
require("vega-embed@6")
)});
  main.variable(observer("raw_nodes_data")).define("raw_nodes_data", function(){return(
[
  {
    cluster_medium: 1,
    unique_id: 111,
    first_name: "Ntaan ",
    surname: "Mcdonald",
    dob: "1996-09-09",
    city: "London",
    email: null,
    group: 29,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 1,
    unique_id: 56,
    first_name: "Nathan ",
    surname: "Mcdonald",
    dob: "1996-09-09",
    city: "London",
    email: "barbara84@robbins.biz",
    group: 29,
    source_dataset: "df_1"
  },
  {
    cluster_medium: 1,
    unique_id: 55,
    first_name: "Nathan ",
    surname: "Mcdonald",
    dob: "1996-09-09",
    city: "London",
    email: "barbara84@robbins.biz",
    group: 29,
    source_dataset: "df_1"
  },
  {
    cluster_medium: 1,
    unique_id: 112,
    first_name: "Nathan ",
    surname: "lconaMd",
    dob: "1996-09-09",
    city: "London",
    email: "birara84@robbins.baz",
    group: 29,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 1,
    unique_id: 109,
    first_name: null,
    surname: "Mcdonald",
    dob: "1996-09-09",
    city: "London",
    email: "barbara84@robbins.biz",
    group: 29,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 2,
    unique_id: 118,
    first_name: "Nancy ",
    surname: "Taylor",
    dob: "1989-07-25",
    city: "London",
    email: "wagnershane@landry.com",
    group: 62,
    source_dataset: "df_1"
  },
  {
    cluster_medium: 2,
    unique_id: 247,
    first_name: "Taylor",
    surname: "Nancy ",
    dob: "1989-07-25",
    city: "London",
    email: "wagnershane@landry.com",
    group: 62,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 2,
    unique_id: 245,
    first_name: "Nancy ",
    surname: "Taylor",
    dob: "1989-07-25",
    city: "London",
    email: "wagnershane@landry.com",
    group: 62,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 2,
    unique_id: 246,
    first_name: "Nancy ",
    surname: "Taylor",
    dob: "1989-08-19",
    city: "London",
    email: "wagnershane@landry.com",
    group: 62,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 2,
    unique_id: 117,
    first_name: null,
    surname: "Taylor",
    dob: "1989-07-25",
    city: "London",
    email: "wagnershane@landry.com",
    group: 62,
    source_dataset: "df_1"
  },
  {
    cluster_medium: 2,
    unique_id: 248,
    first_name: "Nacy ",
    surname: "Taylor",
    dob: "1989-07-25",
    city: "London",
    email: "wagnershane@landry.com",
    group: 62,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 3,
    unique_id: 119,
    first_name: "Lewis",
    surname: "Freddie ",
    dob: "1971-12-19",
    city: "Southend-on-Sea",
    email: "tyler28@weaver-alvarez.com",
    group: 63,
    source_dataset: "df_1"
  },
  {
    cluster_medium: 3,
    unique_id: 120,
    first_name: "Lewis",
    surname: "Freddie ",
    dob: "1971-09-14",
    city: "Southend-on-Sea",
    email: "tyler28@weaver-alvarez.com",
    group: 63,
    source_dataset: "df_1"
  },
  {
    cluster_medium: 3,
    unique_id: 249,
    first_name: "Freddi ",
    surname: "Lewis",
    dob: "1971-12-19",
    city: "Southend-on-Sea",
    email: "tyler28@wecveralvarez.aom",
    group: 63,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 3,
    unique_id: 250,
    first_name: null,
    surname: "Lewis",
    dob: "1971-12-19",
    city: "Southend-on-Sea",
    email: "tyler28@weaver-alvarez.com",
    group: 63,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 4,
    unique_id: 11,
    first_name: "lmeAi ",
    surname: "Alexander",
    dob: "1983-05-19",
    city: "Glasgow",
    email: "icampbell@allen-lewis.org",
    group: 3,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 4,
    unique_id: 10,
    first_name: "Amelia ",
    surname: "Alexander",
    dob: "1983-05-19",
    city: "Glasgow",
    email: "icampbell@allen-lewis.org",
    group: 3,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 6,
    unique_id: 121,
    first_name: "Adam ",
    surname: "Berry",
    dob: "1986-02-08",
    city: "London",
    email: "walterdunn@pham-hill.com",
    group: 64,
    source_dataset: "df_1"
  },
  {
    cluster_medium: 6,
    unique_id: 252,
    first_name: "Adam ",
    surname: "Berr",
    dob: "1986-02-08",
    city: "London",
    email: "walterdunn@pham-hill.com",
    group: 64,
    source_dataset: "df_2"
  },
  {
    cluster_medium: 6,
    unique_id: 253,
    first_name: "Aad ",
    surname: "Berry",
    dob: "1986-02-08",
    city: "London",
    email: "walterdunn@pham-hill.com",
    group: 64,
    source_dataset: "df_2"
  }
]
)});
  main.variable(observer("raw_edges_data")).define("raw_edges_data", function(){return(
[
  {
    tf_adjusted_match_prob: 0.9999999929,
    match_probability: 0.9999917496,
    source_dataset_l: "df_1",
    unique_id_l: 55,
    source_dataset_r: "df_2",
    unique_id_r: 109,
    first_name_l: "Nathan ",
    first_name_r: null,
    gamma_first_name: -1,
    surname_l: "Mcdonald",
    surname_r: "Mcdonald",
    gamma_surname: 2,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: "barbara84@robbins.biz",
    gamma_email: 1,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.9999999929,
    match_probability: 0.9999917496,
    source_dataset_l: "df_1",
    unique_id_l: 56,
    source_dataset_r: "df_2",
    unique_id_r: 109,
    first_name_l: "Nathan ",
    first_name_r: null,
    gamma_first_name: -1,
    surname_l: "Mcdonald",
    surname_r: "Mcdonald",
    gamma_surname: 2,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: "barbara84@robbins.biz",
    gamma_email: 1,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.9996251539,
    match_probability: 0.988646879,
    source_dataset_l: "df_1",
    unique_id_l: 55,
    source_dataset_r: "df_2",
    unique_id_r: 112,
    first_name_l: "Nathan ",
    first_name_r: "Nathan ",
    gamma_first_name: 2,
    surname_l: "Mcdonald",
    surname_r: "lconaMd",
    gamma_surname: 0,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: "birara84@robbins.baz",
    gamma_email: 0,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.9996251539,
    match_probability: 0.988646879,
    source_dataset_l: "df_1",
    unique_id_l: 56,
    source_dataset_r: "df_2",
    unique_id_r: 112,
    first_name_l: "Nathan ",
    first_name_r: "Nathan ",
    gamma_first_name: 2,
    surname_l: "Mcdonald",
    surname_r: "lconaMd",
    gamma_surname: 0,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: "birara84@robbins.baz",
    gamma_email: 0,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.994051709,
    match_probability: 0.994051709,
    source_dataset_l: "df_2",
    unique_id_l: 111,
    source_dataset_r: "df_2",
    unique_id_r: 112,
    first_name_l: "Ntaan ",
    first_name_r: "Nathan ",
    gamma_first_name: 1,
    surname_l: "Mcdonald",
    surname_r: "lconaMd",
    gamma_surname: 0,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: null,
    email_r: "birara84@robbins.baz",
    gamma_email: -1,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.4404287501,
    match_probability: 0.4404287501,
    source_dataset_l: "df_2",
    unique_id_l: 109,
    source_dataset_r: "df_2",
    unique_id_r: 112,
    first_name_l: null,
    first_name_r: "Nathan ",
    gamma_first_name: -1,
    surname_l: "Mcdonald",
    surname_r: "lconaMd",
    gamma_surname: 0,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: "birara84@robbins.baz",
    gamma_email: 0,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.995049863,
    match_probability: 0.995049863,
    source_dataset_l: "df_2",
    unique_id_l: 245,
    source_dataset_r: "df_2",
    unique_id_r: 247,
    first_name_l: "Nancy ",
    first_name_r: "Taylor",
    gamma_first_name: 0,
    surname_l: "Taylor",
    surname_r: "Nancy ",
    gamma_surname: 0,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.2701297195,
    match_probability: 0.2701297195,
    source_dataset_l: "df_2",
    unique_id_l: 246,
    source_dataset_r: "df_2",
    unique_id_r: 247,
    first_name_l: "Nancy ",
    first_name_r: "Taylor",
    gamma_first_name: 0,
    surname_l: "Taylor",
    surname_r: "Nancy ",
    gamma_surname: 0,
    dob_l: "1989-08-19",
    dob_r: "1989-07-25",
    gamma_dob: 0,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.99816319,
    match_probability: 0.99816319,
    source_dataset_l: "df_1",
    unique_id_l: 117,
    source_dataset_r: "df_2",
    unique_id_r: 247,
    first_name_l: null,
    first_name_r: "Taylor",
    gamma_first_name: -1,
    surname_l: "Taylor",
    surname_r: "Nancy ",
    gamma_surname: 0,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.995049863,
    match_probability: 0.995049863,
    source_dataset_l: "df_1",
    unique_id_l: 118,
    source_dataset_r: "df_2",
    unique_id_r: 247,
    first_name_l: "Nancy ",
    first_name_r: "Taylor",
    gamma_first_name: 0,
    surname_l: "Taylor",
    surname_r: "Nancy ",
    gamma_surname: 0,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.5001360677,
    match_probability: 0.5001360677,
    source_dataset_l: "df_1",
    unique_id_l: 120,
    source_dataset_r: "df_2",
    unique_id_r: 250,
    first_name_l: "Lewis",
    first_name_r: null,
    gamma_first_name: -1,
    surname_l: "Freddie ",
    surname_r: "Lewis",
    gamma_surname: 0,
    dob_l: "1971-09-14",
    dob_r: "1971-12-19",
    gamma_dob: 0,
    email_l: "tyler28@weaver-alvarez.com",
    email_r: "tyler28@weaver-alvarez.com",
    gamma_email: 1,
    group_l: 63,
    group_r: 63,
    cluster_medium_l: 3,
    cluster_medium_r: 3
  },
  {
    tf_adjusted_match_prob: 0.9999017555,
    match_probability: 0.994335954,
    source_dataset_l: "df_2",
    unique_id_l: 249,
    source_dataset_r: "df_2",
    unique_id_r: 250,
    first_name_l: "Freddi ",
    first_name_r: null,
    gamma_first_name: -1,
    surname_l: "Lewis",
    surname_r: "Lewis",
    gamma_surname: 2,
    dob_l: "1971-12-19",
    dob_r: "1971-12-19",
    gamma_dob: 1,
    email_l: "tyler28@wecveralvarez.aom",
    email_r: "tyler28@weaver-alvarez.com",
    gamma_email: 0,
    group_l: 63,
    group_r: 63,
    cluster_medium_l: 3,
    cluster_medium_r: 3
  },
  {
    tf_adjusted_match_prob: 0.99816319,
    match_probability: 0.99816319,
    source_dataset_l: "df_1",
    unique_id_l: 119,
    source_dataset_r: "df_2",
    unique_id_r: 250,
    first_name_l: "Lewis",
    first_name_r: null,
    gamma_first_name: -1,
    surname_l: "Freddie ",
    surname_r: "Lewis",
    gamma_surname: 0,
    dob_l: "1971-12-19",
    dob_r: "1971-12-19",
    gamma_dob: 1,
    email_l: "tyler28@weaver-alvarez.com",
    email_r: "tyler28@weaver-alvarez.com",
    gamma_email: 1,
    group_l: 63,
    group_r: 63,
    cluster_medium_l: 3,
    cluster_medium_r: 3
  },
  {
    tf_adjusted_match_prob: 0.9999566321,
    match_probability: 0.9999917496,
    source_dataset_l: "df_1",
    unique_id_l: 117,
    source_dataset_r: "df_1",
    unique_id_r: 118,
    first_name_l: null,
    first_name_r: "Nancy ",
    gamma_first_name: -1,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 1.0,
    match_probability: 0.9999999254,
    source_dataset_l: "df_1",
    unique_id_l: 55,
    source_dataset_r: "df_1",
    unique_id_r: 56,
    first_name_l: "Nathan ",
    first_name_r: "Nathan ",
    gamma_first_name: 2,
    surname_l: "Mcdonald",
    surname_r: "Mcdonald",
    gamma_surname: 2,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: "barbara84@robbins.biz",
    gamma_email: 1,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.9999999769,
    match_probability: 0.9999731722,
    source_dataset_l: "df_1",
    unique_id_l: 55,
    source_dataset_r: "df_2",
    unique_id_r: 111,
    first_name_l: "Nathan ",
    first_name_r: "Ntaan ",
    gamma_first_name: 1,
    surname_l: "Mcdonald",
    surname_r: "Mcdonald",
    gamma_surname: 2,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: null,
    gamma_email: -1,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.9999999769,
    match_probability: 0.9999731722,
    source_dataset_l: "df_1",
    unique_id_l: 56,
    source_dataset_r: "df_2",
    unique_id_r: 111,
    first_name_l: "Nathan ",
    first_name_r: "Ntaan ",
    gamma_first_name: 1,
    surname_l: "Mcdonald",
    surname_r: "Mcdonald",
    gamma_surname: 2,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: null,
    gamma_email: -1,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.9999985892,
    match_probability: 0.9983643831,
    source_dataset_l: "df_2",
    unique_id_l: 109,
    source_dataset_r: "df_2",
    unique_id_r: 111,
    first_name_l: null,
    first_name_r: "Ntaan ",
    gamma_first_name: -1,
    surname_l: "Mcdonald",
    surname_r: "Mcdonald",
    gamma_surname: 2,
    dob_l: "1996-09-09",
    dob_r: "1996-09-09",
    gamma_dob: 1,
    email_l: "barbara84@robbins.biz",
    email_r: null,
    gamma_email: -1,
    group_l: 29,
    group_r: 29,
    cluster_medium_l: 1,
    cluster_medium_r: 1
  },
  {
    tf_adjusted_match_prob: 0.9999999754,
    match_probability: 0.9999595001,
    source_dataset_l: "df_2",
    unique_id_l: 245,
    source_dataset_r: "df_2",
    unique_id_r: 246,
    first_name_l: "Nancy ",
    first_name_r: "Nancy ",
    gamma_first_name: 2,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-08-19",
    gamma_dob: 0,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.9769868124,
    match_probability: 0.9955389635,
    source_dataset_l: "df_1",
    unique_id_l: 117,
    source_dataset_r: "df_2",
    unique_id_r: 246,
    first_name_l: null,
    first_name_r: "Nancy ",
    gamma_first_name: -1,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-08-19",
    gamma_dob: 0,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.9999999754,
    match_probability: 0.9999595001,
    source_dataset_l: "df_1",
    unique_id_l: 118,
    source_dataset_r: "df_2",
    unique_id_r: 246,
    first_name_l: "Nancy ",
    first_name_r: "Nancy ",
    gamma_first_name: 2,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-08-19",
    gamma_dob: 0,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.9999992898,
    match_probability: 0.9999998649,
    source_dataset_l: "df_2",
    unique_id_l: 245,
    source_dataset_r: "df_2",
    unique_id_r: 248,
    first_name_l: "Nancy ",
    first_name_r: "Nacy ",
    gamma_first_name: 1,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.9996144105,
    match_probability: 0.9999266246,
    source_dataset_l: "df_2",
    unique_id_l: 246,
    source_dataset_r: "df_2",
    unique_id_r: 248,
    first_name_l: "Nancy ",
    first_name_r: "Nacy ",
    gamma_first_name: 1,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-08-19",
    dob_r: "1989-07-25",
    gamma_dob: 0,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.995049863,
    match_probability: 0.995049863,
    source_dataset_l: "df_2",
    unique_id_l: 247,
    source_dataset_r: "df_2",
    unique_id_r: 248,
    first_name_l: "Taylor",
    first_name_r: "Nacy ",
    gamma_first_name: 0,
    surname_l: "Nancy ",
    surname_r: "Taylor",
    gamma_surname: 0,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.9999566321,
    match_probability: 0.9999917496,
    source_dataset_l: "df_1",
    unique_id_l: 117,
    source_dataset_r: "df_2",
    unique_id_r: 248,
    first_name_l: null,
    first_name_r: "Nacy ",
    gamma_first_name: -1,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 0.9999992898,
    match_probability: 0.9999998649,
    source_dataset_l: "df_1",
    unique_id_l: 118,
    source_dataset_r: "df_2",
    unique_id_r: 248,
    first_name_l: "Nancy ",
    first_name_r: "Nacy ",
    gamma_first_name: 1,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 1.0,
    match_probability: 0.9999595001,
    source_dataset_l: "df_1",
    unique_id_l: 119,
    source_dataset_r: "df_1",
    unique_id_r: 120,
    first_name_l: "Lewis",
    first_name_r: "Lewis",
    gamma_first_name: 2,
    surname_l: "Freddie ",
    surname_r: "Freddie ",
    gamma_surname: 2,
    dob_l: "1971-12-19",
    dob_r: "1971-09-14",
    gamma_dob: 0,
    email_l: "tyler28@weaver-alvarez.com",
    email_r: "tyler28@weaver-alvarez.com",
    gamma_email: 1,
    group_l: 63,
    group_r: 63,
    cluster_medium_l: 3,
    cluster_medium_r: 3
  },
  {
    tf_adjusted_match_prob: 0.0005357682,
    match_probability: 0.0005357682,
    source_dataset_l: "df_1",
    unique_id_l: 120,
    source_dataset_r: "df_2",
    unique_id_r: 249,
    first_name_l: "Lewis",
    first_name_r: "Freddi ",
    gamma_first_name: 0,
    surname_l: "Freddie ",
    surname_r: "Lewis",
    gamma_surname: 0,
    dob_l: "1971-09-14",
    dob_r: "1971-12-19",
    gamma_dob: 0,
    email_l: "tyler28@weaver-alvarez.com",
    email_r: "tyler28@wecveralvarez.aom",
    gamma_email: 0,
    group_l: 63,
    group_r: 63,
    cluster_medium_l: 3,
    cluster_medium_r: 3
  },
  {
    tf_adjusted_match_prob: 0.2254941519,
    match_probability: 0.2254941519,
    source_dataset_l: "df_1",
    unique_id_l: 119,
    source_dataset_r: "df_2",
    unique_id_r: 249,
    first_name_l: "Lewis",
    first_name_r: "Freddi ",
    gamma_first_name: 0,
    surname_l: "Freddie ",
    surname_r: "Lewis",
    gamma_surname: 0,
    dob_l: "1971-12-19",
    dob_r: "1971-12-19",
    gamma_dob: 1,
    email_l: "tyler28@weaver-alvarez.com",
    email_r: "tyler28@wecveralvarez.aom",
    gamma_email: 0,
    group_l: 63,
    group_r: 63,
    cluster_medium_l: 3,
    cluster_medium_r: 3
  },
  {
    tf_adjusted_match_prob: 0.999970108,
    match_probability: 0.9999776963,
    source_dataset_l: "df_2",
    unique_id_l: 10,
    source_dataset_r: "df_2",
    unique_id_r: 11,
    first_name_l: "Amelia ",
    first_name_r: "lmeAi ",
    gamma_first_name: 0,
    surname_l: "Alexander",
    surname_r: "Alexander",
    gamma_surname: 2,
    dob_l: "1983-05-19",
    dob_r: "1983-05-19",
    gamma_dob: 1,
    email_l: "icampbell@allen-lewis.org",
    email_r: "icampbell@allen-lewis.org",
    gamma_email: 1,
    group_l: 3,
    group_r: 3,
    cluster_medium_l: 4,
    cluster_medium_r: 4
  },
  {
    tf_adjusted_match_prob: 0.999958329,
    match_probability: 0.999958329,
    source_dataset_l: "df_2",
    unique_id_l: 252,
    source_dataset_r: "df_2",
    unique_id_r: 253,
    first_name_l: "Adam ",
    first_name_r: "Aad ",
    gamma_first_name: 0,
    surname_l: "Berr",
    surname_r: "Berry",
    gamma_surname: 1,
    dob_l: "1986-02-08",
    dob_r: "1986-02-08",
    gamma_dob: 1,
    email_l: "walterdunn@pham-hill.com",
    email_r: "walterdunn@pham-hill.com",
    gamma_email: 1,
    group_l: 64,
    group_r: 64,
    cluster_medium_l: 6,
    cluster_medium_r: 6
  },
  {
    tf_adjusted_match_prob: 0.9999999985,
    match_probability: 0.9999776963,
    source_dataset_l: "df_1",
    unique_id_l: 121,
    source_dataset_r: "df_2",
    unique_id_r: 253,
    first_name_l: "Adam ",
    first_name_r: "Aad ",
    gamma_first_name: 0,
    surname_l: "Berry",
    surname_r: "Berry",
    gamma_surname: 2,
    dob_l: "1986-02-08",
    dob_r: "1986-02-08",
    gamma_dob: 1,
    email_l: "walterdunn@pham-hill.com",
    email_r: "walterdunn@pham-hill.com",
    gamma_email: 1,
    group_l: 64,
    group_r: 64,
    cluster_medium_l: 6,
    cluster_medium_r: 6
  },
  {
    tf_adjusted_match_prob: 0.9999566321,
    match_probability: 0.9999917496,
    source_dataset_l: "df_1",
    unique_id_l: 117,
    source_dataset_r: "df_2",
    unique_id_r: 245,
    first_name_l: null,
    first_name_r: "Nancy ",
    gamma_first_name: -1,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 1.0,
    match_probability: 0.9999999254,
    source_dataset_l: "df_1",
    unique_id_l: 118,
    source_dataset_r: "df_2",
    unique_id_r: 245,
    first_name_l: "Nancy ",
    first_name_r: "Nancy ",
    gamma_first_name: 2,
    surname_l: "Taylor",
    surname_r: "Taylor",
    gamma_surname: 2,
    dob_l: "1989-07-25",
    dob_r: "1989-07-25",
    gamma_dob: 1,
    email_l: "wagnershane@landry.com",
    email_r: "wagnershane@landry.com",
    gamma_email: 1,
    group_l: 62,
    group_r: 62,
    cluster_medium_l: 2,
    cluster_medium_r: 2
  },
  {
    tf_adjusted_match_prob: 1.0,
    match_probability: 0.9999998607,
    source_dataset_l: "df_1",
    unique_id_l: 121,
    source_dataset_r: "df_2",
    unique_id_r: 252,
    first_name_l: "Adam ",
    first_name_r: "Adam ",
    gamma_first_name: 2,
    surname_l: "Berry",
    surname_r: "Berr",
    gamma_surname: 1,
    dob_l: "1986-02-08",
    dob_r: "1986-02-08",
    gamma_dob: 1,
    email_l: "walterdunn@pham-hill.com",
    email_r: "walterdunn@pham-hill.com",
    gamma_email: 1,
    group_l: 64,
    group_r: 64,
    cluster_medium_l: 6,
    cluster_medium_r: 6
  }
]
)});
  main.variable(observer("splink_settings")).define("splink_settings", function(){return(
{
  link_type: "link_and_dedupe",
  blocking_rules: ["l.city = r.city"],
  comparison_columns: [
    {
      col_name: "first_name",
      num_levels: 3,
      term_frequency_adjustments: true,
      gamma_index: 0,
      data_type: "string",
      fix_u_probabilities: false,
      fix_m_probabilities: false,
      case_expression:
        "case\n    when (first_name_l is null or first_name_r is null) then -1\n    when jaro_winkler_sim(first_name_l, first_name_r) >= 1.0 then 2\n    when jaro_winkler_sim(first_name_l, first_name_r) >= 0.88 then 1\n    else 0 end as gamma_first_name",
      m_probabilities: [
        0.36731645464897156,
        0.17439667880535126,
        0.4582868814468384
      ],
      u_probabilities: [
        0.9930018782615662,
        0.0028558988124132156,
        0.0041422066278755665
      ]
    },
    {
      col_name: "surname",
      num_levels: 3,
      term_frequency_adjustments: true,
      gamma_index: 1,
      data_type: "string",
      fix_u_probabilities: false,
      fix_m_probabilities: false,
      case_expression:
        "case\n    when (surname_l is null or surname_r is null) then -1\n    when jaro_winkler_sim(surname_l, surname_r) >= 1.0 then 2\n    when jaro_winkler_sim(surname_l, surname_r) >= 0.88 then 1\n    else 0 end as gamma_surname",
      m_probabilities: [
        0.34741583466529846,
        0.12415848672389984,
        0.5284256935119629
      ],
      u_probabilities: [
        0.9902822375297546,
        0.0029645985923707485,
        0.006753162480890751
      ]
    },
    {
      col_name: "dob",
      gamma_index: 2,
      num_levels: 2,
      data_type: "string",
      term_frequency_adjustments: false,
      fix_u_probabilities: false,
      fix_m_probabilities: false,
      case_expression:
        "case\n    when dob_l is null or dob_r is null then -1\n    when dob_l = dob_r then 1\n    else 0 end as gamma_dob",
      m_probabilities: [0.3700525462627411, 0.6299474835395813],
      u_probabilities: [0.996875524520874, 0.0031245029531419277]
    },
    {
      col_name: "email",
      gamma_index: 3,
      num_levels: 2,
      data_type: "string",
      term_frequency_adjustments: false,
      fix_u_probabilities: false,
      fix_m_probabilities: false,
      case_expression:
        "case\n    when email_l is null or email_r is null then -1\n    when email_l = email_r then 1\n    else 0 end as gamma_email",
      m_probabilities: [0.2865733206272125, 0.7134267091751099],
      u_probabilities: [0.9964072108268738, 0.00359280314296484]
    }
  ],
  additional_columns_to_retain: ["group"],
  em_convergence: 0.0001,
  source_dataset_column_name: "source_dataset",
  unique_id_column_name: "unique_id",
  retain_matching_columns: true,
  retain_intermediate_calculation_columns: false,
  max_iterations: 25,
  proportion_of_matches: 0.03724955394864082
}
)});
  main.variable(observer("svu_options")).define("svu_options", function(){return(
{
  cluster_colname: 'cluster_medium'
}
)});
  return main;
}
